modules/amm-executor/src/main/scala/org/ergoplatform/dex/executor/amm/services/CFMMBacklog.scala

```diff
@@ -0,0 +1,202 @@
+    } yield order
+
+    /** Put an order from the backlog.
+      */
+    def drop(id: OrderId): F[Unit] =
+      cfmmOrders.drop(id)
+
+    def getMaxOrderFromQueue(queue: Atom[F, mutable.PriorityQueue[WeightedOrder]]): F[Option[CFMMOrder]] =
+      for {
+        maxId <- queue.modify { queue =>
+                   val elem = if (queue.isEmpty) none else queue.dequeue().some
+                   (queue, elem)
+                 }
+        time <- now.millis
+        suspendedElem <- maxId match {
+                           case Some(value) if (time - value.timestamp) < backlogConfig.orderLifetime.toMillis =>
+                             cfmmOrders.get(value.orderId) >>= {
+                               case Some(value) => value.some.pure[F]
+                               case None        => getMaxOrderFromQueue(queue)
+                             }
+                           case Some(value) =>
+                             cfmmOrders.drop(value.orderId) >> getMaxOrderFromQueue(queue)
+                           case None => none.pure[F]
+                         }
+      } yield suspendedElem
+
+    def filterRevisit: F[Unit] = for {
+      curTime <- now.millis
+      possible2pending <-
+        revisitOrders.modify(_.span(wOrd => (curTime - wOrd.timestamp > backlogConfig.orderExecutionTime.toMillis)))
+      _ <- possible2pending.traverse {
+             case ord if (curTime - ord.timestamp) > backlogConfig.orderLifetime.toMillis =>
+               drop(ord.orderId)
+             case ord =>
+               pendingQueueRef.update(_ += ord)
+           }
+    } yield ()
+  }
+
+  final private class CFMMBacklogTracingMid[F[_]: FlatMap: Logging] extends CFMMBacklog[Mid[F, *]] {
+
+    /** Put an order to the backlog.
+      */
+    override def put(order: CFMMOrder): Mid[F, Unit] = for {
+      _ <- trace"put(order=$order)"
+      r <- _
+      _ <- trace"put(order=$order) -> $r"
+    } yield r
+
+    /** Put an order with priceTooHigh or priceTooLow execution err
+      */
+    override def suspend(order: CFMMOrder): Mid[F, Unit] = for {
+      _ <- trace"suspend(order=$order)"
+      r <- _
+      _ <- trace"suspend(order=$order) -> $r"
+    } yield r
+
+    /** Put possibly executed order to the backlog
+      */
+    override def checkLater(order: CFMMOrder): Mid[F, Unit] = for {
+      _ <- trace"checkLater(order=$order)"
+      r <- _
+      _ <- trace"checkLater(order=$order) -> $r"
+    } yield r
+
+    /** Get candidate order for execution. Blocks until an order is available.
+      */
+    override def get: Mid[F, Option[CFMMOrder]] = for {
+      _ <- trace"get()"
+      r <- _
+      _ <- trace"get() -> $r"
+    } yield r
+
+    /** Put an order from the backlog.
+      */
+    override def drop(id: OrderId): Mid[F, Unit] = for {
+      _ <- trace"drop(id=$id)"
+      r <- _
+      _ <- trace"drop(id=$id) -> $r"
+    } yield r
+  }
+
+  private def recoverPendingQueue[F[_]: Monad: Clock: BacklogConfig.Has](
```

Comment:
I think it's much better to implement this using streaming api.

Put it please above the Trace impl.

Reply:
why better?

Reply:
I think, this queue possibly can hold millions of orders and the app memory won't hold it and app will crash with oom error